cpp/demo/poisson_noufc/main.cpp

```diff
@@ -0,0 +1,591 @@
+//
+// The Poisson equation is the canonical elliptic partial differential
+// equation.  For a domain :math:`\Omega \subset \mathbb{R}^n` with
+// boundary :math:`\partial \Omega = \Gamma_{D} \cup \Gamma_{N}`, the
+// Poisson equation with particular boundary conditions reads:
+//
+// .. math::
+//    - \nabla^{2} u &= f \quad {\rm in} \ \Omega, \\
+//      u &= 0 \quad {\rm on} \ \Gamma_{D}, \\
+//      \nabla u \cdot n &= g \quad {\rm on} \ \Gamma_{N}. \\
+//
+// Here, :math:`f` and :math:`g` are input data and :math:`n` denotes the
+// outward directed boundary normal. The most standard variational form
+// of Poisson equation reads: find :math:`u \in V` such that
+//
+// .. math::
+//    a(u, v) = L(v) \quad \forall \ v \in V,
+//
+// where :math:`V` is a suitable function space and
+//
+// .. math::
+//    a(u, v) &= \int_{\Omega} \nabla u \cdot \nabla v \, {\rm d} x, \\
+//    L(v)    &= \int_{\Omega} f v \, {\rm d} x
+//    + \int_{\Gamma_{N}} g v \, {\rm d} s.
+//
+// The expression :math:`a(u, v)` is the bilinear form and :math:`L(v)`
+// is the linear form. It is assumed that all functions in :math:`V`
+// satisfy the Dirichlet boundary conditions (:math:`u = 0 \ {\rm on} \
+// \Gamma_{D}`).
+//
+// In this demo, we shall consider the following definitions of the input
+// functions, the domain, and the boundaries:
+//
+// * :math:`\Omega = [0,1] \times [0,1]` (a unit square)
+// * :math:`\Gamma_{D} = \{(0, y) \cup (1, y) \subset \partial \Omega\}`
+// (Dirichlet boundary)
+// * :math:`\Gamma_{N} = \{(x, 0) \cup (x, 1) \subset \partial \Omega\}`
+// (Neumann boundary)
+// * :math:`g = \sin(5x)` (normal derivative)
+// * :math:`f = 10\exp(-((x - 0.5)^2 + (y - 0.5)^2) / 0.02)` (source term)
+//
+//
+// Implementation
+// --------------
+//
+// The implementation is split in two files: a form file containing the
+// definition of the variational forms expressed in UFL and a C++ file
+// containing the actual solver.
+//
+// Running this demo requires the files: :download:`main.cpp`,
+// :download:`Poisson.ufl` and :download:`CMakeLists.txt`.
+//
+//
+// UFL form file
+// ^^^^^^^^^^^^^
+//
+// The UFL file is implemented in :download:`Poisson.ufl`, and the
+// explanation of the UFL file can be found at :doc:`here <Poisson.ufl>`.
+//
+//
+// C++ program
+// ^^^^^^^^^^^
+//
+// The main solver is implemented in the :download:`main.cpp` file.
+//
+// At the top we include the DOLFIN header file and the generated header
+// file "Poisson.h" containing the variational forms for the Poisson
+// equation.  For convenience we also include the DOLFIN namespace.
+//
+// .. code-block:: cpp
+
+#include <Eigen/Dense>
+#include <cfloat>
+#include <dolfin.h>
+#include <dolfin/mesh/Ordering.h>
+
+using namespace dolfin;
+
+// Then follows the definition of the coefficient functions (for
+// :math:`f` and :math:`g`), which are derived from the
+// :cpp:class:`Expression` class in DOLFIN
+//
+// .. code-block:: cpp
+
+class P1Element
```

Comment:
I don't see the need for a class. Could use a namespace if you want to group functions.